src/supernova/mod.rs

```diff
@@ -129,139 +113,226 @@ where
+)]
+pub struct AuxParams<G1, G2>
 where
   G1: Group<Base = <G2 as Group>::Scalar>,
   G2: Group<Base = <G1 as Group>::Scalar>,
-  C1: StepCircuit<G1::Scalar>,
-  C2: StepCircuit<G2::Scalar>,
 {
-  /// Indexed `RunningClaim`s
-  claims: Vec<RunningClaim<G1, G2, C1, C2>>,
-  /// Digest constructed from these `RunningClaim`s' parameters
-  // Once we serialize this struct, it's important to add this
-  // annotaiton to the digest field for it to be treated properly: #[serde(skip, default = "OnceCell::new")]
-  digest: OnceCell<G1::Scalar>,
+  ck_primary: CommitmentKey<G1>, // This is shared between all circuit params
+
+  F_arity_secondary: usize,
+  ro_consts_secondary: ROConstants<G2>,
+  ro_consts_circuit_secondary: ROConstantsCircuit<G1>,
+  ck_secondary: CommitmentKey<G2>,
+  r1cs_shape_secondary: R1CSShape<G2>,
+  augmented_circuit_params_secondary: SuperNovaAugmentedCircuitParams,
+
+  #[abomonate_with(<G1::Scalar as PrimeField>::Repr)]
+  digest: G1::Scalar,
 }
 
-impl<G1, G2, C1, C2> Index<usize> for RunningClaims<G1, G2, C1, C2>
+impl<G1, G2, C1, C2> Index<usize> for PublicParams<G1, G2, C1, C2>
 where
   G1: Group<Base = <G2 as Group>::Scalar>,
   G2: Group<Base = <G1 as Group>::Scalar>,
-  C1: StepCircuit<G1::Scalar>,
-  C2: StepCircuit<G2::Scalar>,
+  C1: EnforcingStepCircuit<G1::Scalar>,
+  C2: EnforcingStepCircuit<G2::Scalar>,
 {
-  type Output = RunningClaim<G1, G2, C1, C2>;
+  type Output = CircuitParams<G1, G2>;
 
   fn index(&self, index: usize) -> &Self::Output {
-    &self.claims[index]
+    &self.circuit_params[index]
   }
 }
 
-impl<G1, G2, C1, C2> Digestible for RunningClaims<G1, G2, C1, C2>
+impl<G1, G2, C1, C2> Digestible for PublicParams<G1, G2, C1, C2>
 where
   G1: Group<Base = <G2 as Group>::Scalar>,
   G2: Group<Base = <G1 as Group>::Scalar>,
-  C1: StepCircuit<G1::Scalar>,
-  C2: StepCircuit<G2::Scalar>,
+  C1: EnforcingStepCircuit<G1::Scalar>,
+  C2: EnforcingStepCircuit<G2::Scalar>,
 {
   fn write_bytes<W: Sized + io::Write>(&self, byte_sink: &mut W) -> Result<(), io::Error> {
-    assert!(!self.claims.is_empty());
-    for claim in &self.claims {
-      claim.get_public_params().write_bytes(byte_sink)?;
+    assert!(!self.circuit_params.is_empty());
+    for pp in &self.circuit_params {
+      pp.write_bytes(byte_sink)?;
     }
     Ok(())
   }
 }
 
-impl<G1, G2, C1, C2> RunningClaims<G1, G2, C1, C2>
+impl<G1, G2, C1, C2> PublicParams<G1, G2, C1, C2>
 where
   G1: Group<Base = <G2 as Group>::Scalar>,
   G2: Group<Base = <G1 as Group>::Scalar>,
-  C1: StepCircuit<G1::Scalar>,
-  C2: StepCircuit<G2::Scalar>,
+  C1: EnforcingStepCircuit<G1::Scalar>,
+  C2: EnforcingStepCircuit<G2::Scalar>,
 {
-  fn new(mut claims: Vec<RunningClaim<G1, G2, C1, C2>>) -> Self {
-    let running_claim_params = claims
+  /// new running claim params
+  pub fn new<NC: NonUniformCircuit<G1, G2, C1, C2>>(
+    non_unifrom_circuit: &NC,
+  ) -> Self {
+    // uhhh if this aint zero this doesn't work
```

Comment:
Does the output value of `iter_base_step` clear things up, and if so, could we refer to it here?